src/core/io/src/4C_io_input_spec_validators.hpp

```diff
@@ -30,78 +31,202 @@ FOUR_C_NAMESPACE_OPEN
-      Predicate pred;
-      Describe desc;
-      EmitMetadata emit;
+      using type = T;
+    };
 
-      bool operator()(const T& v) const { return pred(v); }
-      void describe(std::ostream& os) const { desc(os); }
-      void emit_metadata(YamlNodeRef node) const { emit(node); }
+    template <typename T>
+    struct GeneralizedType<RangeLike<T>>
+    {
+      using type = RangeLike<typename GeneralizedType<T>::type>;
     };
 
-    Concept impl_;
-
-   public:
-    //! Type-erasing constructor.
-    template <typename FPredicate, typename FDescribe, typename FEmitMetadata>
-    Validator(FPredicate p, FDescribe d, FEmitMetadata em)
-        : impl_{
-              .pred = Predicate(std::move(p)),
-              .desc = Describe(std::move(d)),
-              .emit = EmitMetadata(std::move(em)),
-          }
+    template <typename T>
+    struct GeneralizedType<std::vector<T>>
     {
-    }
+      using type = RangeLike<typename GeneralizedType<T>::type>;
+    };
 
-    /**
-     * The main validation function. Returns true if the value is valid according to the
-     * predicate stored inside this validator. Returns false otherwise.
-     *
-     * @note This operator can only be called with a value that is exactly of type T without any
-     * implicit conversions.
-     */
-    template <typename U>
-      requires(std::same_as<T, std::decay_t<U>>)
-    [[nodiscard]] bool operator()(const U& v) const
+    template <typename T, std::size_t N>
+    struct GeneralizedType<std::array<T, N>>
     {
-      return impl_(v);
-    }
+      using type = RangeLike<typename GeneralizedType<T>::type>;
+    };
 
-    /**
-     * Describes what this validator expects from the value in human-readable form .
-     */
-    void describe(std::ostream& os) const { impl_.describe(os); }
+    template <typename T>
+    struct GeneralizedType<std::optional<T>>
+    {
+      using type = std::optional<typename GeneralizedType<T>::type>;
+    };
 
-    /**
-     * Emits metadata about this validator into the given YAML node.
-     */
-    void emit_metadata(YamlNodeRef node) const { impl_.emit_metadata(node); }
+    template <typename T>
+    struct GeneralizedType<std::map<std::string, T>>
+    {
+      using type = std::map<std::string, typename GeneralizedType<T>::type>;
+    };
 
-    friend std::ostream& operator<<(std::ostream& os, const Validator& v)
+    template <typename T1, typename T2>
+    struct GeneralizedType<std::pair<T1, T2>>
     {
-      v.describe(os);
-      return os;
-    }
-  };
+      using type =
+          std::pair<typename GeneralizedType<T1>::type, typename GeneralizedType<T2>::type>;
+    };
+
+    template <typename... Ts>
+    struct GeneralizedType<std::tuple<Ts...>>
+    {
+      using type = std::tuple<typename GeneralizedType<Ts>::type...>;
+    };
+
+    template <typename T>
+    using GeneralizedTypeT = typename GeneralizedType<T>::type;
+
+
+    template <typename T, typename Enable = void>
+    class ValidatorImpl
```

Comment:
`Validator` is specialized for different cases. I moved this into `Internal` and now call it `ValidatorImpl`.